### PR TITLE
Read Google Drive directories concurrently

### PR DESCRIPTION
**Pull request by klauspost**
This PR will read Google drive directories concurrently.

Maximum connection handling is moved to the "pacer", and removed from ACD.

It seems like root drive listing is handled elsewhere, but that performs really well, so I have not done anything to change that.

### CODE REVIEW

File: pacer/pacer.go [hunk 1]
```diff
@@ -91,6 +111,9 @@ func (p *Pacer) beginCall() {
 	// Ticker more accurately, but then we'd have to work out how
 	// not to run it when it wasn't needed
 	<-p.pacer
+	if p.maxConnections > 0 {
```
Comment (klauspost) [inline pacer/pacer.go]: This is on purpose not inside the mutex lock, and this is also one of the reasons of the "do not change MaxConnections" after calling the pacer.

The "if" check could be put inside a lock/unlock, The `<-p.connTokens` and `p.connTokens <- struct{}{}` may not hold the mutex. But if we change that and allow for MaxConnections while requests are running, we got even bigger problems with "token overload". So unless we really need to change "MaxConnections" on the fly this is the cleanest solution.

**Comment by klauspost** [comment]
This fixes #143 if used with "-v".  We could still add a more generic progress display for file listing.

File: pacer/pacer.go (hunk 1 above)
Comment (ncw) [inline pacer/pacer.go]: Yes I think that works perfectly.

**Comment by ncw** [comment]
Putting the concurrency control inside the pacer is a great idea.

I did a bit of testing with drive and it speeds my directory listings up by a factor of 3.5!

There is a way of listing everything on drive so we use that for the root directory listing.

Here is a comparison of the root directory listing
```
$ rclone ls drive: > 0

Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:       19.1s

$ rclone --drive-full-list=false ls drive: > 1

Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:        7.5s

$ diff <(sort <0) <(sort <1)
```

Whether it is faster or slower will depend on your drive contents I think...


I've merged that - thanks!

**Comment by ncw** [comment]
Thinking about this futher, it is short selling the concurrency in the sync phase slightly which is supposed to be config.Transfers and cofig.Checkers....  So the default could be config.Transfers+config.Checkers which over sells the concurrency in the listing phase, but I'd be happy with that.

**Pull request closed without merging**